File: src/cody/grblgui/Simulation.scala
package cody.grblgui

import Array.ofDim

class Simulation(size_x : Float, size_y : Float, size_z : Float, precision : Float) {
	val count_x : Int = (size_x / precision).toInt;
	val count_y : Int = (size_y / precision).toInt;
	var map:Array[Array[Float]] = ofDim[Float](count_x,count_y)
	var zero = Vector2D(-size_x / 2f,-size_y / 2f)
	
	def index_to_position(x : Int, y : Int) : Vector2D = Vector2D(zero.x + x * precision, zero.y + y * precision)
	
	def position_to_index(v : Vector2D) = (((v.x - zero.x) / precision).toInt, ((v.y - zero.y) / precision).toInt)
	def position_to_index(x : Float, y : Float) = (((x - zero.x) / precision).toInt, ((y - zero.y) / precision).toInt)
	def length_to_count(l : Float) : Int = (l / precision).toInt
	
	def getZ(x : Int, y : Int) = map(x)(y)
	def getZminmax() : (Float, Float) = {
	  var min : Float = Float.MaxValue
	  var max : Float = Float.MinValue
	  for(x <- 0 to count_x - 1) {
		  for(y <- 0 to count_y - 1) {
		    val z = getZ(x,y)
		    min = math.min(z, min)
		    max = math.max(z, max)
		  }
	  }
	  return (min, max)
	}
	
	def mill(x : Int, y : Int, z : Float) {
	  if(x >= 0 && y >= 0 && x < count_x && y < count_y)
		  map(x)(y) = math.min(map(x)(y), z)
	}
	
	def mill(pos : Vector3D, tool : ToolInfo) {
	  val center = position_to_index(pos.x, pos.y)
	  val count = length_to_count(tool.getRadius())
	  for(x <- center._1 - count to center._1 + count) {
		  for(y <- center._2 - count to center._2 + count) {
			  var p = index_to_position(x,y)
			  p = Vector2D(p.x - pos.x, p.y - pos.y)
			  
			  val dist = p.length
			  if(dist < tool.getRadius()) {
			    //println("mill " + pos.x + " " + pos.y + " " + pos.z + " " + x + " " + y)
			    mill(x, y, pos.z)
			  }
		  }
	  }
	}
	
	def simulate(path : Toolpath, tool : ToolInfo) {
	  var cur = path.path.get(0)
	  for(i <- 1 to path.path.size() - 1) {
	    val next = path.path.get(i)
	    if(next.z < 0 || cur.z < 0){
	      var d = next.cpy()
	      d.sub(cur)
	      //println("line " + cur.toString() + " -> " + next.toString())
	      val segments = math.max(1,(d.len() / precision).toInt)
	      for(j <- 0 to segments - 1) {
	        var p = d.cpy()
	        p.mul(j)
	        p.div(segments)
	        p.add(cur)
	        mill(Vector3D(p.x,p.y,p.z), tool)
	      }
	    }
	    cur = next
	  }
	}
	
	
}

File: grblgui/src/cody/grblgui/Simulation.java
package cody.grblgui;

import com.badlogic.gdx.math.GridPoint2;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;

public class Simulation {
	public int count_x, count_y;
	float map[][];
	Vector2 zero;
	final float precision;
	
	public Simulation(float size_x, float size_y, float size_z, float precision) {
		this.precision = precision;
		count_x = (int)(size_x / precision);
		count_y = (int)(size_y / precision);
		
		map = new float[count_x][count_y];
		zero = new Vector2(-size_x / 2f,-size_y / 2f);
	}
	
	public Vector2 indexToPosition(int x, int y) {
		return new Vector2(zero.x + x * precision, zero.y + y * precision);
	}
	
	public GridPoint2 positionToIndex(Vector2 v) {
		return new GridPoint2((int)((v.x - zero.x) / precision), (int)((v.y - zero.y) / precision));
	}
	
	public GridPoint2 positionToIndex(float x, float y) {
		return new GridPoint2((int)((x - zero.x) / precision), (int)((y - zero.y) / precision));
	}
	
	public int lengthToCount(float l) {
		return (int)(l/precision);
	}
	
	public float getZ(int x, int y) {
		return map[x][y];
	}
	
	public Vector2 getZminmax() {
		float min = Float.MAX_VALUE;
		float max = Float.MIN_VALUE;
		  for(int x = 0; x <= count_x - 1; ++x) {
			  for(int y = 0; y <= count_y - 1; ++y) {
			    float z = getZ(x,y);
			    min = Math.min(z, min);
			    max = Math.max(z, max);
			  }
		  }
		  return new Vector2(min, max);
	}
	
	void mill(int x, int y, float z) {
		if(x >= 0 && y >= 0 && x < count_x && y < count_y)
			map[x][y] = Math.min(map[x][y], z);
	}
	
	void mill(Vector3 pos, ToolInfo tool) {
		GridPoint2 center = positionToIndex(pos.x, pos.y);
	  int count = lengthToCount(tool.getRadius());
	  
	  for(int x = center.x - count; x <= center.x + count; ++x) {
		  for(int y = center.y - count; y <= center.y + count; ++y) {
			  Vector2 p = indexToPosition(x,y);
			  p = new Vector2(p.x - pos.x, p.y - pos.y);
			  
			  float dist = p.len();
			  if(dist < tool.getRadius()) {
			    //println("mill " + pos.x + " " + pos.y + " " + pos.z + " " + x + " " + y)
			    mill(x, y, pos.z);
			  }
		  }
	  }
	}
	
	void simulate(Toolpath path, ToolInfo tool) {
		Vector3 cur = path.path.get(0);
		
		for(int i=1;i<=path.path.size() - 1;++i) {
				    Vector3 next = path.path.get(i);
				    if(next.z < 0 || cur.z < 0){
				      Vector3 d = next.cpy();
				      d.sub(cur);
				      //println("line " + cur.toString() + " -> " + next.toString())
				      int segments = Math.max(1,(int)(d.len() / precision));
				      for(int j = 0; j <= segments - 1; ++j) {
				        Vector3 p = d.cpy();
				        p.scl(j);
				        p.scl(1.0f/segments);
				        p.add(cur);
				        mill(new Vector3(p.x,p.y,p.z), tool);
				      }
				    }
				    cur = next;
				  }
	}
}
